runtime/src/bank_forks.rs

```diff
@@ -378,4 +390,74 @@ mod tests {
         bank_forks.insert(child_bank);
         assert_eq!(bank_forks.active_banks(), vec![1]);
     }
+
+    #[test]
+    fn test_bank_forks_different_set_root() {
+        solana_logger::setup();
+        let leader_keypair = Keypair::new();
+        let GenesisConfigInfo {
+            mut genesis_config,
+            mint_keypair: _,
+            voting_keypair,
+        } = create_genesis_config_with_leader(10_000, &leader_keypair.pubkey(), 1_000);
+        let slots_in_epoch = 32;
+        genesis_config.epoch_schedule = EpochSchedule::new(slots_in_epoch);
+
+        let bank0 = Bank::new(&genesis_config);
+        let mut bank_forks0 = BankForks::new(bank0);
+
+        let bank1 = Bank::new(&genesis_config);
+        let mut bank_forks1 = BankForks::new(bank1);
+        bank_forks0.set_root(0, &None, None);
+
+        let additional_timestamp_secs = 2;
+
+        let num_slots = slots_in_epoch + DEPRECATED_TIMESTAMP_SLOT_RANGE as u64 + 1;
+        for slot in 1..num_slots {
+            // Just after the epoch boundary, timestamp a vote that will shift
+            // Clock::unix_timestamp from Bank::unix_timestamp_from_genesis()
+            let update_timestamp_case = slot == slots_in_epoch;
+
+            let child1 = Bank::new_from_parent(&bank_forks0[slot - 1], &Pubkey::default(), slot);
+            if update_timestamp_case {
+                let recent_timestamp: UnixTimestamp = child1.unix_timestamp_from_genesis();
+                update_vote_account_timestamp(
+                    BlockTimestamp {
+                        slot: child1.slot(),
+                        timestamp: recent_timestamp + additional_timestamp_secs,
+                    },
+                    &child1,
+                    &voting_keypair.pubkey(),
+                );
+            }
+            bank_forks0.insert(child1);
+            bank_forks0.set_root(slot, &None, None);
+
+            let child2 = Bank::new_from_parent(&bank_forks1[slot - 1], &Pubkey::default(), slot);
+            if update_timestamp_case {
+                let recent_timestamp: UnixTimestamp = child2.unix_timestamp_from_genesis();
+                update_vote_account_timestamp(
+                    BlockTimestamp {
+                        slot: child2.slot(),
+                        timestamp: recent_timestamp + additional_timestamp_secs,
+                    },
+                    &child2,
+                    &voting_keypair.pubkey(),
+                );
+            }
+            bank_forks1.insert(child2);
+        }
+        let child1 =
+            Bank::new_from_parent(&bank_forks0[num_slots - 1], &Pubkey::default(), num_slots);
+        let child2 =
+            Bank::new_from_parent(&bank_forks1[num_slots - 1], &Pubkey::default(), num_slots);
+        bank_forks1.set_root(num_slots - 1, &None, None);
+
```

Comment:
instead of these new children, can we just reuse the last children that were constructed in the loop?

Reply:
Done https://github.com/solana-labs/solana/pull/13477/commits/1721cb69af4c9f50fa98d638a7a9a15883914395